src/Transaction.php

```diff
@@ -0,0 +1,655 @@
+                    || $message->isAccountSigner($accountIndex);
+                $isWritable = $message->isAccountWritable($accountIndex);
+                return new AccountMeta($publicKey, $isSigner, $isWritable);
+            }, $instruction->accounts);
+
+            array_push($transaction->instructions, new TransactionInstruction(
+                $message->accountKeys[$instruction->programIdIndex],
+                $keys,
+                Ed25519Keypair::bin2array($instruction->data)
+            ));
+        }
+
+        return $transaction;
+    }
+
+    /**
+     * @param array<AccountMeta> $haystack
+     * @param PublicKey|SignaturePubkeyPair|AccountMeta|string $needle
+     * @return int|string
+     */
+    static protected function arraySearchAccountMetaForPublicKey(array $haystack, $needle)
+    {
+        $publicKeyToSearchFor = static::toPublicKey($needle);
+
+        foreach ($haystack as $i => $item) {
+            if (static::toPublicKey($item) == $publicKeyToSearchFor) {
+                return $i;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * @param array $haystack
+     * @return array
+     * @throws GenericException
+     */
+    static protected function arrayUnique(array $haystack)
+    {
+        $unique = [];
+        foreach ($haystack as $item) {
+            $indexOfSigner = static::arraySearchAccountMetaForPublicKey($unique, $item);
+
+            if ($indexOfSigner === -1) {
+                array_push($unique, $item);
+            }
+        }
+
+        return $unique;
+    }
+
+    /**
+     * @param $source
+     * @return PublicKey
+     * @throws GenericException
+     */
+    static protected function toPublicKey($source): PublicKey
+    {
+        if ($source instanceof PublicKey) {
+            return $source;
+        } elseif ($source instanceof SignaturePubkeyPair) {
+            return $source->publicKey;
+        } elseif ($source instanceof AccountMeta) {
+            return $source->publicKey;
+        } elseif (is_string($source)) {
+            return new PublicKey($source);
+        } elseif ($source instanceof KeyPair) {
+            return $source->getPublicKey();
+        } elseif ($source instanceof Signer) {
+            return $source->publicKey;
+        } else {
+            throw new GenericException('Invalid $needle input into arraySearchAccountMetaForPublicKey');
+        }
+    }
+
+    /**
+     * @param $source
+     * @return PublicKey
+     * @throws GenericException
+     */
+    protected function toSecretKey($source): string
+    {
+        if ($source instanceof KeyPair) {
+            return Ed25519Keypair::array2bin($source->getSecretKey());
+        } elseif ($source instanceof Signer) {
+            return $source->secretKey;
+        } else {
+            throw new GenericException('Invalid $needle input into arraySearchAccountMetaForPublicKey');
+        }
+    }
+}
```

Comment:
Again.. glad this class was you and not me. I'm so impressed with how much work you've put into this!